lib/ecto/repo/schema.ex

```diff
@@ -865,29 +865,34 @@ defmodule Ecto.Repo.Schema do
   end
 
   defp pop_assocs(changeset, []) do
-    {changeset, [], []}
+    {changeset, [], [], []}
   end
 
   defp pop_assocs(%{changes: changes, types: types} = changeset, assocs) do
-    {changes, parent, child} =
-      Enum.reduce assocs, {changes, [], []}, fn assoc, {changes, parent, child} ->
+    {changes, parent, child, unchanged_parent} =
+      Enum.reduce assocs, {changes, [], [], []}, fn assoc, {changes, parent, child, unchanged_parent} ->
         case changes do
           %{^assoc => value} ->
             changes = Map.delete(changes, assoc)
 
             case types do
               %{^assoc => {:assoc, %{relationship: :parent} = refl}} ->
-                {changes, [{refl, value} | parent], child}
+                {changes, [{refl, value} | parent], child, unchanged_parent}
               %{^assoc => {:assoc, %{relationship: :child} = refl}} ->
-                {changes, parent, [{refl, value} | child]}
+                {changes, parent, [{refl, value} | child], unchanged_parent}
             end
 
           %{} ->
-            {changes, parent, child}
+            case types do
+              %{^assoc => {:assoc, %{relationship: :parent} = refl}} ->
+                {changes, parent, child, [refl | unchanged_parent]}
```

Comment:
Is it worth already computing if there is a mismatch here? Because we have changes and data as well, right? So we could directly return the list of fields to reset?

Reply:
ah good point. WDYT?